### ISSUE TITLE
Extend std::fs::Permissions on Windows

### ISSUE THREAD

**Issue by E-Mans-Application** [rust-lang/libs-team#388]
# Proposal

## Problem statement

I need to set file attributes on Windows, and I found it's exactly what `std::fs::set_permissions` does. However, struct `std::fs::Permissions` only exposes attribute `readonly`. I would also need to set other attributes, such as `FILE_ATTRIBUTE_SYSTEM` or `FILE_ATTRIBUTE_TEMPORARY`.
On Linux, there is extension trait `std::os::unix::fs::PermissionsExt`, but it seems Windows has no such thing.

The attributes given to `OpenOptions` are only applied to existing files if the files are open with `.create(true).truncate(true)`, which truncates the files. I need to get and set the attributes of an existing file, without changing its contents.

## Solution sketch

An extension trait could be added on Windows to expose other attributes. This trait could also provide methods to convert `Permissions` from and to `u32`.

In `std::sys::pal::windows::fs`, extend `impl FilePermissions`:

```rust
impl FilePermissions {
    pub fn readonly(&self) -> bool {
        self.attrs & c::FILE_ATTRIBUTE_READONLY != 0
    }

    pub fn set_readonly(&mut self, readonly: bool) {
        if readonly {
            self.attrs |= c::FILE_ATTRIBUTE_READONLY;
        } else {
            self.attrs &= !c::FILE_ATTRIBUTE_READONLY;
        }
    }

    pub fn system(&self) -> bool {
        self.attrs & c::FILE_ATTRIBUTE_SYSTEM != 0
    }

    pub fn set_system(&mut self, system: bool) {
        if readonly {
            self.attrs |= c::FILE_ATTRIBUTE_SYSTEM;
        } else {
            self.attrs &= !c::FILE_ATTRIBUTE_SYSTEM;
        }
    }

    // ...
   // A macro could be used since all the functions will have the same structure

    pub fn to_u32(&self) -> u32 {
        self.attrs
    }
    fn from_u32(mask: u32) -> Self {
         // You may want to check the mask is correct and return an `Option` here, although `SetFileAttributes` seems to ignore invalid attributes.
         Self { attrs: mask }
    }
}
```

In `std::os::windows::fs`:
```rust
pub trait PermissionsExt {
    #[must_use]
    fn system(&self) -> bool;
    fn set_system(&mut self);

    #[must_use]
    fn temporary(&self) -> bool;
    fn set_temporary(&mut self);

   // Same for other attributes

    fn to_u32(&self) -> u32;
    fn from_u32(mask: u32) -> Self;
}

impl PermissionsExt for Permissions {
    // Delegates all to `as_inner` or `from_inner`, just like `PermissionsExt` is implemented on Unix.
}
```

## Alternatives

Currently, I'm calling system API `SetFileAttributesW` directly, but this forces me to resort to unsafe code and to re-implement conversion from `Path` to `Vec<u16>` (with verbatim support for long paths), in a way that's probably less efficient than the standard library.

Another possibility would be to compute the mask manually and to transmute the result to Permissions (which is basically a u32), but I don't like this solution because there is no guarantee that Permissions will remain equivalent to a u32 in the future.

## Links and related work

Initial discussion:
https://internals.rust-lang.org/t/extend-std-permissions-on-windows/20943

**Comment by ChrisDenton** [comment]
I'm not the biggest fan of call attributes "permissions" but std does already so maybe that ship has sailed. Though I guess a possible alternative would be to have windows only `FileAttributes` struct.

**Comment by pitaj** [comment]
Yeah calling these all permissions feels wrong. `readonly` feels a lot more like a permission than those other two attributes.

What about a `std::os::windows::fs::FileExt`?

```rust
trait FileExt {
    fn attributes(&mut self) -> FileAttributes;
}
impl FileExt for File {
    ...
}
```

**Comment by ChrisDenton** [comment]
> What about a `std::os::windows::fs::FileExt`?

Another alternative is to simplify this further and just have

```rust
trait FileExt {
    fn attributes(&self) -> u32;
    fn set_attributes(&mut self, attrs: u32);
}
```

This is less ergonomic of course but it more or less matches [`OpenOptionsExt::attributes`](https://doc.rust-lang.org/stable/std/os/windows/fs/trait.OpenOptionsExt.html#tymethod.attributes) and side-steps quibbling over which attributes are important enough to have in std.

**Comment by E-Mans-Application** [comment]
I agree that `permissions` is not a very appropriate name to refer to attributes.

I suggested to extend `Permissions` because, despite its name, it exactly represents the file's attributes. (The unique field of `fs_impl::FilePermissions` is named `attrs` for a reason.)

I don't think `fn attributes` should belong to `File` because it only contains a `HANDLE`, so this would require an additional system call to get the attributes.
However, `fn attributes` could be cheaply added to `Metadata` instead, alongside `Metadata::permissions` (which would return the same field...)

`fn set_attributes` could still be added to `File`, alongside `set_permissions`.

That would result in two functions doing the same thing, but this would work as well.

**Comment by ChrisDenton** [comment]
That sounds good to me.

**Comment by ChrisDenton** [comment]
This was discussed in today's libs-api meeting. The decision was to accept the original proposal to extend `Permissions` (using `PermissionsExt`) because it already does set attributes and this could lead to conflicts where setting readonly effectively undoes other attribute changes.

There was also discussion about naming. It was decided that we should use an explicit name for the to/from `u32` functions so it's more clear what it's doing. E.g.:

```rust
pub trait PermissionsExt: Sealed {
    fn file_attributes(&self) -> u32;
    fn set_file_attributes(mask: u32) -> Self;
}
```

There wasn't a conclusion on which specific attributes should or shouldn't have a special method but that can be decided on a case-by-case basis before stabilization.

**Comment by asder8215** [comment]
I didn't see a tracking issue for this or any implementation of a `PermissionExt` trait for Windows. I hope you don't mind me working on this @E-Mans-Application.

From the discussion here, I'm assuming this `PermissionsExt` trait is a mirror of Unix's `PermissionsExt` where in `library/std/src/os/unix/fs.rs` it's defined here as:
```rust
#[stable(feature = "fs_ext", since = "1.1.0")]
pub trait PermissionsExt {
    /// Returns the mode permission bits
    #[stable(feature = "fs_ext", since = "1.1.0")]
    fn mode(&self) -> u32;

    /// Sets the mode permission bits.
    #[stable(feature = "fs_ext", since = "1.1.0")]
    fn set_mode(&mut self, mode: u32);

    /// Creates a new instance from the given mode permission bits.
    #[stable(feature = "fs_ext", since = "1.1.0")]
    #[cfg_attr(not(test), rustc_diagnostic_item = "permissions_from_mode")]
    fn from_mode(mode: u32) -> Self;
}

#[stable(feature = "fs_ext", since = "1.1.0")]
impl PermissionsExt for Permissions {
    fn mode(&self) -> u32 {
        self.as_inner().mode()
    }

    fn set_mode(&mut self, mode: u32) {
        *self = Permissions::from_inner(FromInner::from_inner(mode));
    }

    fn from_mode(mode: u32) -> Permissions {
        Permissions::from_inner(FromInner::from_inner(mode))
    }
}
```
I haven't explored too much of Windows' file attributes since I use Linux, but I took a look at the file attribute constants that Windows has [here](https://learn.microsoft.com/en-us/windows/win32/fileio/file-attribute-constants). Right now, I'm thinking of implementing a private method for each of the Windows' file attribute in `FilePermissions` within `library/std/src/sys/fs/windows.rs` (using the file attribute flags seen in `library/std/src/sys/pal/windows/c/windows_sys.rs`) like how E-Mans-Application mentioned in his solution sketch with:
```rust
impl FilePermissions {
    pub fn readonly(&self) -> bool {
        self.attrs & c::FILE_ATTRIBUTE_READONLY != 0
    }

    pub fn set_readonly(&mut self, readonly: bool) {
        if readonly {
            self.attrs |= c::FILE_ATTRIBUTE_READONLY;
        } else {
            self.attrs &= !c::FILE_ATTRIBUTE_READONLY;
        }
    }

    fn hidden(&self) -> bool {
        self.attrs & c::FILE_ATTRIBUTE_HIDDEN != 0
    }

    fn set_hidden(&self, hidden: bool) {
        if hidden {
            self.attrs |= c::FILE_ATTRIBUTE_HIDDEN;
        } else {
            self.attrs &= !c::FILE_ATTRIBUTE_HIDDEN;
        }
    }

    fn system(&self) -> bool {
        self.attrs & c::FILE_ATTRIBUTE_SYSTEM != 0
    }

    fn set_system(&self) -> bool {
        if system {
            self.attrs |= c::FILE_ATTRIBUTE_SYSTEM;
        } else {
            self.attrs &= !c::FILE_ATTRIBUTE_SYSTEM;
        }
    }
...
}
```
I'll keep the implementation PR open for if any of these file attribute methods should be public/kept private or removed.

**Comment by asder8215** [comment]
Forgot to mention it, but a tracking issue was created for this, and an implementation for this has been merged in.
